velox/common/memory/HashStringAllocator.h

```diff
@@ -555,29 +555,45 @@ struct AlignedStlAllocator {
   }
 
   T* FOLLY_NONNULL allocate(std::size_t n) {
-    if (n * sizeof(T) > HashStringAllocator::kMaxAlloc && poolAligned_) {
-      return reinterpret_cast<T*>(allocator_->allocateFromPool(n * sizeof(T)));
+    if (n * sizeof(T) > HashStringAllocator::kMaxAlloc) {
+      if (poolAligned_) {
+        return reinterpret_cast<T*>(
+            allocator_->allocateFromPool(n * sizeof(T)));
+      } else {
+        // Memory layout:
+        // - Header
+        // - padding(length is stored in `delta`)
+        // - delta(4 bytes storing the size of padding)
+        // - the aligned ptr
+        auto paddedSize =
+            checkedPlus<size_t>(Alignment + 4, checkedMultiply(n, sizeof(T)));
+        // Allocate the memory block with a Header.
+        auto header = reinterpret_cast<HashStringAllocator::Header*>(
```

Comment:
You don't need to create a header when allocate directly from pool

Reply:
Since HashStringAllocator::free can automatically handle a large allocation prefixed by a Header(see comments below), it seems simpler to allocate a Header with it?

Reply:
We just need to be symmetric here.  Either `allocate` + `free`, or `allocateFromPool` + `freeFromPool` should work, but don't mix `allocateFromPool` + `free` it's little confusing.